### DISCUSSION TITLE
Extend atom onMount with getter

### DISCUSSION THREAD

**Discussion by SomeCodingUser** [pmndrs/jotai#2236] [Ideas] [answered]
Hi,

I was wondering if it would be possible to extend the onMount function with a getter? Currently it is only possible to set a value in there, but it would help improve our code base in some instances.

E.g. currently we registered atom effects for several atoms, in order to do so we need to mount the effect in component.
With a getter we could move this to the onMount function like this:

```export const myValueAtom = atom<string>('');

const myValueAtomEffect = atomEffect((get, set) => {
  const myValue= get(myValueAtom);

  // do something...
});  

myValueAtom.onMount((get) => { // <-- added getter
  get(myValueAtomEffect);
}); ```

If this was possible I would love to get my hands dirty on that feature.

Thanks!

**Comment by dai-shi** [comment, marked as answer]
It's our design choice to only support `set` in onMount. You should create a new atom. We prefer atom composition.

```js
const countAtom = atom(1)
const anotherCountAtom = atom(2)
const initAtom = atom(null, (get, set) => {
  set(anotherCountAtom, get(countAtom) * 10)
})
initAtom.onMount = (init) => {
  init()
}
```

**Reply by SomeCodingUser** [reply]
After thinking about it, is the initAtom.onMount even called if it's not used anywhere else? I assume the onMount function is only called if initAtom is used in a component right?

**Reply by dai-shi** [reply]
Well, we actually need more composition in practice:

```js
export const anotherCountAtom = atom(2)
const baseCountAtom = atom(1)
const initAtom = atom(null, (get, set) => {
  set(anotherCountAtom, get(countAtom) * 10)
})
initAtom.onMount = (init) => {
  init()
}
export const countAtom = atom(
  (get) => {
    get(initAtom)
    return get(baseCountAtom)
  },
  (get, set, ...args) => set(baseCountAtom, ...args)
)
```

Something like that. It depends on what you would like to do.

**Reply by h3adache** [reply]
Semi related to @SomeCodingUser's comment but if `anotherCountAtom` has an `onMount` (for example, I'm using `atomWithStorage` with `getOnInit: true`), this won't trigger it. Is there a reason why `get(xxxAtom)` doesn't also invoke `onMount`?

**Reply by dai-shi** [reply]
There are two kinds of `get`. One is declarative, `get` in `read`. The other is imperative, `get` in `write`. We only mount atoms with declarative `get`. onMount is imperative.

**Reply by h3adache** [reply]
This causes some confusion/pain when we are trying to initialize a set of atoms (large object pre initialization) when there are other atoms with `onMount` but am trying to rework the flow based on declarative gets. Thanks!

**Reply by dmaskasky** [reply]
Does [`atomWithLazy`](https://jotai.org/docs/utilities/lazy) work for you use-case?

**Comment by SomeCodingUser** [comment]
Awesome, thanks!

**Comment by dmaskasky** [comment]
I'm late to the party but the way to do this is with atom composition.

```ts
import { atomEffect } from 'jotai-effect'

const myValueBaseAtom = atom(0)

const myValueAtomEffect = atomEffect((get, set) => {
  const myValue= get(myValueBaseAtom)
})

const myValueAtom = atom(
  (get) => { 
    get(myValueAtomEffect)
    return get(myValueBaseAtom)
  },
  (_get, set, value: number) => set(myValueBaseAtom, value)
)
```

Alternatively, you can just use withAtomEffect, which is equivalent to above.
```ts
import { withAtomEffect } from 'jotai-effect'

const myValueAtom = withAtomEffect(atom(0), (get, set) => {
  const myValue= get(myValueAtom)
})
```
